### DISCUSSION TITLE
Will the window operation state survive a process restart?

### DISCUSSION THREAD

**Discussion by vol-ale** [lf-edge/ekuiper#2129] [Q&A] [answered]
Hello guy's,

Thank you for the fantastic product! 

According to the documentation, eKuiper supports [Internal state for window operation and rewindable source](https://ekuiper.org/docs/en/latest/guide/rules/state_and_fault_tolerance.html). 
Let's assume I have a MQTT source with QOS 1 and a rule with a Counting Window of 5. If during receiving the 3rd message, the eKuiper process will be restarted, will it automatically replay the first two messages to conform to the "at least once" delivery, or I need to configure an external storage for this? Assuming that these first two messages were check-pointed. Is MQTT source rewindable?

**Comment by ngjaying** [comment, marked as answer]
Thank you @vol-ale 

The state storage is sqlite by default. You can configure it in etc/kuiper.yaml, please check https://ekuiper.org/docs/en/latest/configuration/global_configurations.html#store-configurations.

MQTT source is not rewindable.

**Reply by vol-ale** [reply]
@ngjaying, and so, do I understand correctly that if I have the MQTT source, the checkpointed rule state will be restored, but messages won't be replayed because of MQTT protocol "nature"?

**Reply by ngjaying** [reply]
Correct.

**Reply by vol-ale** [reply]
Thanks for a quick response!
